zephyr/drivers/can/can_sccan.c

```diff
@@ -0,0 +1,1147 @@
+	}
+
+	if (sc_can_is_tx_fifo_full(config)) {
+		return -EAGAIN;
+	}
+
+	ret = k_mutex_lock(&data->inst_mutex, SCCAN_MUTEX_LOCK_TIMEOUT);
+	if (ret != 0) {
+		return -ETIMEDOUT;
+	}
+
+	sc_can_set_idr(config, frame);
+
+	sc_can_set_dlc(config, frame->dlc);
+
+	sc_can_set_data_frame(config, frame);
+
+	/* Save call back function */
+	/* TODO:
+	 * In the current FPA CAN IP core, CAN Packets stored in the TX Buffer
+	 * transmit to the CAN Bus in a FIFO (First In First Out).
+	 * Therefore, there is no issue with managing the list of TX callbacks
+	 * using an array.
+	 * However, due to future modifications in the FPGA CAN IP core, the
+	 * transmission order of CAN Packets stored in the TX buffer will be
+	 * rearranged based on the CAN ID (priority).
+	 * So, we will need to change the management of TX callbacks from an
+	 * array to something like an slist.
+	 */
+	data->tx_cb_data_list[data->tx_head].tx_cb = callback;
+	data->tx_cb_data_list[data->tx_head].tx_cb_arg = user_data;
+	data->tx_head++;
+	if (data->tx_head == config->tx_fifo_depth) {
+		data->tx_head = 0;
+	}
+
+	k_mutex_unlock(&data->inst_mutex);
+
+	return 0;
+}
+
+/*
+ * SC CAN controller has four "Acceptance filter" feature, so this driver use
+ * it as the "RX filter".
+ * And, due to the specifications of the SC CAN controller, Acceptance Filter
+ * can only be set when disable CAN.
+ * However, Zephyr specification don't defines the timing for adding/removing
+ * RX filter. In fact, CAN sample application (samples/drivers/can/counter) is
+ * called can_add_rx_filter_msgq() after can_start().
+ * Therefore, in this driver, temporarily disabled CAN and set
+ * "Acceptance filer" to SC CAN controller.
+ */
+static int sc_can_add_rx_filter(const struct device *dev, can_rx_callback_t cb, void *cb_arg,
+				const struct can_filter *filter)
+{
+	const struct sc_can_cfg *config = dev->config;
+	struct sc_can_data *data = dev->data;
+	bool need_enable = false;
+	int ret;
+	int filter_id;
+
+	LOG_DBG("Setting filter ID: 0x%x, mask: 0x%x", filter->id, filter->mask);
+
+	ret = k_mutex_lock(&data->inst_mutex, SCCAN_MUTEX_LOCK_TIMEOUT);
+	if (ret != 0) {
+		return -ETIMEDOUT;
+	}
+
+	if (sc_can_is_enabled(config)) {
+		/* Disable CAN with retry if enabled */
+		ret = sc_can_disable(config);
+		if (ret != 0) {
+			goto unlock;
+		}
+		need_enable = true;
+	}
+
+	ret = -ENOSPC;
+	for (filter_id = 0; filter_id < config->max_filter; filter_id++) {
+		if (!sc_can_filter_is_used(config, filter_id)) {
+			sc_can_add_acceptance_filter(config, filter_id, filter);
+
+			data->rx_cb_data_list[filter_id].rx_cb = cb;
+			data->rx_cb_data_list[filter_id].rx_cb_arg = cb_arg;
+			data->rx_filter_list[filter_id] = *filter;
+			ret = filter_id;
+			break;
+		}
+	}
+
+	if (filter_id == config->max_filter) {
+		LOG_ERR("No free filter left");
+	} else {
+		LOG_DBG("Filter added. ID: %d", filter_id);
+	}
+
+	if (need_enable) {
+		/* Enable CAN if disabled in this API */
+		sc_can_enable(config);
```

Comment:
Not checking the return value.  There are other call sights as well.

Reply:
Will fix it.